### ISSUE TITLE
It's not useful to send callback stream IDs + cudaStreamAddCallback() deprecation

### ISSUE THREAD

**Issue by eyalroz** [eyalroz/cuda-api-wrappers#103]
We live in a C++ world. A callback can easily have a reference to its stream if it wants to:
```
my_stream.enqueue.callback(
            [&my_stream](cuda::status_t status) { /* do stuff */ }
);
```
plus, a stream ID doesn't even tell you what device the stream is on, so it's not super-useful - and you might not even be able to determine the device, since CUDA Runtime API calls are mostly forbidden  from within callbacks; and to get the device ID differently would mean having callbacks that take `stream_t`'s.

So either we make callbacks take stream_t's rather than stream IDs, or we make them no stream ID, just the status.

Also... in CUDA 10.0, `cudaStreamAddCallback()` is deprecated in favor of `cudaLaunchHostFunc()` - and the host function doesn't get a stream ID not a status (and otherwise the semantics seem to be the same). So maybe what we really need to do is drop _both_ extra parameters?

@codecircuit : What do you guys think?

**Comment by codecircuit** [comment]
> So maybe what we really need to do is drop both extra parameters?

So you think the lambda should not take any arguments at all, and capture all the relevant values?

**Comment by eyalroz** [comment]
@codecircuit : I would have liked to keep the status value and not the stream ID which is not that useful. But nVIDIA is telling us we won't be getting that value for much longer. So, bottom line, I guess so.

**Comment by codecircuit** [comment]
It seems reasonable to me to give a lambda with one argument of type `cuda::status_t`, if the users wants to use the callback API. Just as you suggested it in this issue.

If `cuda-api-wrappers` provides a wrapper for `cudaLaunchHostFunc` it seems most reasonable to give a lambda with no arguments at all.

**Comment by eyalroz** [comment]
@codecircuit : The thing is, I'm not thrilled to offer two callback enqueue APIs, one for functions which take a status and one for functions which don't - even if I hack some template voodoo to distinguish the two cases. So my plan is to use cudaStreamAddCallback() for `CUDA < 10` and `cudaLaunchHostFunc()` for CUDA >= 10. That would mean no `status_t`.

**Comment by codecircuit** [comment]
Ah, okay. I thought you want to offer two different APIs. One for `cudaLaunchHostFunc` and one for `cudaStreamAddCallback`. 

What happens if you use `CUDA < 10`  and the function given to `cudaStreamAddCallback` gets a status type which is `!= cudaSuccess`? Do you throw an error in that case? Because the user won't be able to handle the CUDA status value in the lambda function any more.

**Comment by eyalroz** [comment]
No, it'll just need to get the status from CUDA. Which it can't :-(  ...  because the callback mechanism is totally broken IMHO. They _should_ have allowed making any non-blocking API call from within a callback.

**Comment by codecircuit** [comment]
Why do you think it is totally broken? What non-blocking API call within a callback would be useful?

**Comment by eyalroz** [comment]
@codecircuit : You want to be able to use them for managing/scheduling future work on the GPU. After some kernel or D2H transfer has concluded, you have additional information regarding what to do next (both its execution time and specific results which you've gotten back) - then you want to execute some more work accordingly. They way your host code gets notified about this happening is either blocking on a stream, or via callbacks. But with a callback you can't do anything CUDA-related, so you have to tell a _different_ thread to do it. Which means a different thread needs to be waiting anyways, or you need to launch a new thread. Ugh.

**Comment by codecircuit** [comment]
I agree. In case of workloads with dynamic problem sizes it would be useful to call CUDA functions within the callback.

**Comment by eyalroz** [comment]
I opened a bug about this 2 years ago:
https://developer.nvidia.com/nvidia_bug/2016444

... no comment from nVIDIA so far.

**Comment by eyalroz** [comment]
Oh, and by the way - in OpenCL this is supported, in the sense that you're only forbidden from using blocking OpenCL calls in a callback.

**Comment by eyalroz** [comment]
Fixed long ago by commit 106cc40!
